### PR TITLE
StatefulSet Upgrade Test

### PR DESCRIPTION
**Pull request by kow3ns**
Adds StatefulSet upgrade tests and moves common functionality into the framework package. This removes the potential for cyclic dependencies while allowing for code reuse.
```release-note
NONE
```

### CODE REVIEW

**Comment by kow3ns** [comment]
@k8s-bot test this

**Comment by kow3ns** [comment]
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221)*

File: test/e2e/framework/statefulset_utils.go [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	. "github.com/onsi/gomega"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
```
Comment (sttts) [inline test/e2e/framework/statefulset_utils.go]: new line

File: test/e2e/framework/statefulset_utils.go [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+
+// VerifyPodAtIndex applies a visitor patter to the Pod at index in ss. verify is is applied to the Pod to "visit" it.
+func (s *StatefulSetTester) VerifyPodAtIndex(index int, ss *apps.StatefulSet, verify VerifyStatefulPodFunc) {
+	name := getPodNameAtIndex(index, ss)
+	pod, err := s.c.Core().Pods(ss.Namespace).Get(name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Failed to get stateful pod %s for StatefulSet %s/%s", name, ss.Namespace, ss.Name))
+	verify(pod)
+}
+
+func getPodNameAtIndex(index int, ss *apps.StatefulSet) string {
```
Comment (sttts) [inline test/e2e/framework/statefulset_utils.go]: put some StatefulSet into the name

File: test/e2e/framework/statefulset_utils.go [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+	if pollErr != nil {
+		errList = append(errList, fmt.Sprintf("Timeout waiting for pv provisioner to delete pvs, this might mean the test leaked pvs."))
+	}
+	if len(errList) != 0 {
+		ExpectNoError(fmt.Errorf("%v", strings.Join(errList, "\n")))
+	}
+}
+
+// Returns true if pod's StatefulSetInitAnnotation exists and is set to true.
+func IsInitialized(pod v1.Pod) bool {
```
Comment (sttts) [inline test/e2e/framework/statefulset_utils.go]: put some StatefulSet into the name

File: test/e2e/framework/statefulset_utils.go [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+				},
+			},
+			VolumeClaimTemplates: claims,
+			ServiceName:          governingSvcName,
+		},
+	}
+}
+
+// SetInitializedAnnotation sets teh StatefulSetInitAnnotation to value.
+func SetInitializedAnnotation(ss *apps.StatefulSet, value string) {
```
Comment (sttts) [inline test/e2e/framework/statefulset_utils.go]: put some StatefulSet into the name

**Comment by sttts** [comment]
Looks good. Just some naming issues.

**Comment by kow3ns** [comment]
Reviewed 5 of 6 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[test/e2e/framework/statefulset_utils.go, line 28 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221#-Kc_AxTCyof-u-KHhEhn-r1-28:-KcctBoQbdC84-xFDtRN:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c2c7f74c403321d3bd1132eda9eb3b103506fc46/test/e2e/framework/statefulset_utils.go#L28)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

new line
</blockquote></details>

Done.

---

*[test/e2e/framework/statefulset_utils.go, line 192 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221#-Kc_AxTCyof-u-KHhEhn-r1-192:-Kcctd-VbjqOYA8FY0v5:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c2c7f74c403321d3bd1132eda9eb3b103506fc46/test/e2e/framework/statefulset_utils.go#L192)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

put some StatefulSet into the name
</blockquote></details>

Done.

---

*[test/e2e/framework/statefulset_utils.go, line 463 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221#-Kc_AxTCyof-u-KHhEhn-r1-463:-Kcctds847_hmCkLAXfn:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c2c7f74c403321d3bd1132eda9eb3b103506fc46/test/e2e/framework/statefulset_utils.go#L463)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

put some StatefulSet into the name
</blockquote></details>

Done.

---

*[test/e2e/framework/statefulset_utils.go, line 556 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221#-Kc_AxTCyof-u-KHhEhn-r1-556:-Kccv_OpBpygXlLuKjio:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c2c7f74c403321d3bd1132eda9eb3b103506fc46/test/e2e/framework/statefulset_utils.go#L556)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

put some StatefulSet into the name
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221)*

File: test/e2e/upgrades/statefulset.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+	By("Creating statefulset " + ssName + " in namespace " + ns)
+	*(t.set.Spec.Replicas) = 3
+	_, err = f.ClientSet.Apps().StatefulSets(ns).Create(t.set)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Saturating stateful set " + t.set.Name)
+	t.tester.Saturate(t.set)
+
+	By("Verifying statefulset mounted data directory is usable")
```
Comment (krousey) [inline test/e2e/upgrades/statefulset.go]: This and subsequent tests are repeated verbatim in the `Test` method. You could put them in a private method that both `Setup` and `Test` call.

**Review by krousey — CHANGES_REQUESTED** [review]

**Comment by kow3ns** [comment]
Reviewed 2 of 2 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e/upgrades/statefulset.go, line 65 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221#-Kc_AxTCyof-u-KHhEho-r2-65:-KcdFCrkd-EanOrpMGdZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/88a677f1909eeb69d32d5ac1c3bb9859f6ad7001/test/e2e/upgrades/statefulset.go#L65)):*
<details><summary><i>Previously, krousey wrote…</i></summary><blockquote>

This and subsequent tests are repeated verbatim in the `Test` method. You could put them in a private method that both `Setup` and `Test` call.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221)*

**Comment by kow3ns** [comment]
Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41221)*

**Comment by krousey** [comment]
I'm not a big fan of moving everything into test/e2e/framework. I would much rather see something like test/e2e/util/statefulsetutil/ (and others).

**Comment by kow3ns** [comment]
@k8s-bot gci gke e2e test this

**Comment by kow3ns** [comment]
"I'm not a big fan of moving everything into test/e2e/framework. I would much rather see something like test/e2e/util/statefulsetutil/ (and others)."

I concur, but that is the way its currently being done for Pods, Services, Deployments, Nodes, Federation, ..., and the canonical example for upgrade tests is supposed to be Services. In this instance, I think that consistency, for the sake of maintenance, is the right approach. Perhaps we should have a larger discussion about the package structure of e2e in the context of go package naming and organization best practices.

**Review by krousey — APPROVED** [review]

**Comment by krousey** [comment]
/approve

I'll let @sttts have the final lgtm.

**Comment by kow3ns** [comment]
@stts are there any further comments that need to be addressed.

**Pull request merged**